### DISCUSSION TITLE
API breaking changes, backward compatibility and legacy methods

### DISCUSSION THREAD

**Discussion by m-holger** [qpdf/qpdf-dev#15] [General]
From #1 : 

>  In some cases, QPDFObjectHandle exposes way too much stuff, some of which better belongs to helpers, but much of it predates helpers

> Also, some of the helpers maintain caches. While my intention was to have them created ephemerally, used for a local change, and destroyed, this is fairly inefficient.

>  The qpdf API is already bloated and inconsistent.

> I also think it may be time to work toward a pivot where we break what needs to be broken and fix some old mistakes in the API. ...  and removing things in the qpdf API that would be better served in other ways with more modern versions of the C++ standard library.

>  I have had a rigorous adherence to backward compatibility

I feel that the backward compatibility is one of the strength of qpdf / qpdflib, and adding major API breaking changes may be detrimental as users may decide to abandon qpdf  rather than update their programs to make the necessary changes. On the other hand, maintaining backward compatibility severely ties our hands.

One possible solution to the conundrum would be to make such change wherever possible on an opt-in basis. For example, we could have a static method `QPDF::API(x, y)` and if `(x, y) >= (11, 10)` we could repair the pages tree after reading the xref table, treat uninitialised object handles as null, etc.

The main advantages of this approach would be:

- we can add additional validations and repairs without having to worry whether they are API breaking for some obscure use of qpdf.

- we can add improvements to the qpdf application at each release rather than having to wait  for a major version changes (eg the JSON duplicate key change). 

I am also wondering if we should start marking methods that are provided for backward compatibility as legacy, with a build option to treat them as deprecated.

**Comment by jberkenbilt** [comment]
I like both the opt-in idea and the marking of methods. I wish I had been marking methods when when they were introduced, but I haven't formally done that. All that said, there are details to work out. When you start doing feature flagging, you can descend into a treacherous space of having 2^n ways to use the library, at which point it becomes much harder to meaningfully test various feature combinations, and it also risks further complicating an already complex API. There are many more options for PDF than there used to be, and qpdf has a decent legacy user base, so certain types of improvements may annoy more users than they will help. I was looking at the landscape of go and rust PDF libraries, and both languages seem to have some decent stuff.

**Reply by m-holger** [reply]
> When you start doing feature flagging, you can descend into a treacherous space of having 2^n ways to use the library,

I was not thinking of allowing a pick and mix approach - opting into API(12, 3) would mean opting into all of API(12,3) as well as opting into all of API(12,2), API(12,1), ..., API(11, 10). This reduces it to n ways to use the library. As in the majority of cases the API change will mean that calling old methods becomes a logic error, for most of the detailed tests this should reduce to 2 ways.

**Reply by jberkenbilt** [reply]
I guess I'd have to see an actual example of this to be convinced that it would be worth the complexity. I think something like this could be good, but we have to use it sparingly and make sure it's not needed for basic uses.

**Comment by m-holger** [comment]
One API breaking change that I would like to make would be to require library users to declare if they are using the library for investigative purposes or to create invalid/broken files. 

If users do not declare that they are using qpdf for investigative purposes, the result of modifying a QPDF object in a way that does not comply with the PDF spec will be undefined, meaning that the behaviour may change from version to version without being considered an API change. In particular, we may choose in the future to silently repair, warn and repair, or throw an exception. There should be a commitment to document any such change in behaviour in the release notes.

If users declare that they are using qpdf for investigative purposes, qpdf will continue to accept changes that do not comply with the PDF spec (although, as already is the case, higher level functions don't work if their pre-conditions are not met).

**Reply by jberkenbilt** [reply]
I agree with this. Using qpdf to intentionally create broken files is going to be a rare use case, and people who are doing that are going to be able to handle a higher level of complexity. As far as automatically repairing files, we should only do that when we're sure we're not going to make anything worse. Repair of the page tree is an example of where there's a high level of certainty, and repairing the xref table is a case where we have no choice and usually (but not always) get it right, but qpdf should "first do no harm," and I think that means not repairing things we can accommodate in most cases. For the most part, this has kept qpdf fully future compatible as well. The only change I remember having to make to qpdf to allow it to accept newer files was introducing UTF-8 strings for the PDF 2.0 spec. When I wrote qpdf initially, PDF 1.3 was current, so the change to `#` in tokens was already there.

So...mostly this should apply to API-created objects and constructs, not to things already in the file, and I would think this would mostly apply to higher-level APIs.

**Reply by m-holger** [reply]
We are in agreement re those points.
